classes/logger.php

```diff
@@ -0,0 +1,180 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * MFA management class.
+ *
+ * @package     tool_heartbeat
+ * @author      Srdjan Janković <srdjan@catalyst.net.nz>
+ * @copyright   Catalyst IT
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_heartbeat;
+
+use core_shutdown_manager;
+use moodle_exception, Exception;
+
+defined('MOODLE_INTERNAL') || die();
+
+class logger {
+    private static $classname = __NAMESPACE__.'\logger';
+    protected static $statcallbacks = [];
+
+    /**
+     * Register shutdown handler
+     *
+     * @return void
+     */
+    public static function register_shutdown_handler(): void {
+        core_shutdown_manager::register_function([self::$classname, 'log']);
+    }
+
+    /**
+     * Register log callback
+     *
+     * @param callable $cb
+     * @param ?string  $stream A fopen() compliant string
+     * @return void
+     */
+    public static function register_stat_callback(callable $cb, $stream = null): void {
+        if (!isset(self::$statcallbacks[$stream])) {
+            self::$statcallbacks[$stream] = [];
+        }
+        self::$statcallbacks[$stream][] = $cb;
+    }
+
+    /**
+     * Shutdown handler
+     *
+     * @return void
+     */
+    public static function log(): void {
+        $defaultstream = self::default_log_stream();
+        $statcallbacks = self::$statcallbacks;
```

Comment:
It isn't very clear what all the lines from here to line 80 are doing. All of the $statcallbacks stuff seems very complicated, I believe it is all to allow injection for the unit tests? Is so could this just be tucked away into a getter?